### DISCUSSION TITLE
"No jobs available!" due to lack of pages for long deltas

### DISCUSSION THREAD

**Discussion by TheFearCode** [jrouwe/JoltPhysics#959] [Q&A] [answered]
Not sure if I'm doing somethiing wrong but, when I have huge delta time like 1 second or more the engine simply stalls due to lack of static memory to allocate more jobs.
this happens at:
`system.Update(dt, collisionSteps, &_allocator, &_jobSystem);`
when collisionSteps >= 60 (assuming the game is 60 FPS, so that's a one second lag).
The lag is caused basically when I debug network features, I could simply override it and it works, but I wonder if that's expected because I could simply override (lying the steps when debugging) or just statically rebuilding Jolt with 4096 as max jobs.

**Comment by jankrassnigg** [comment]
You are definitely doing something wrong giving the engine such huge delta timesteps. Call it repeatedly with a timestep of 1/60 until the simulation caught up. Otherwise the simulation stability will be awful anyway.

**Reply by TheFearCode** [reply]
I see, so what's the point of the collisionSteps if I have to repeatedly call Update with 1 step?
Also won't it kill the simulation's parallelism if I call it repeatedly instead of increasing `collisionSteps`?
Maybe a hybrid approach would be better?

**Reply by jankrassnigg** [reply, marked as answer]
To my understanding the collision step option allows you to let the engine do one part of the simulation with higher accuracy. Usually you would hard-code this to something like 4. Then for every simulation tick it does 4 sub-steps of one part of the simulation, which gives better results without taking the full 4x simulation time.

Generally these physics engines are designed for real-time use cases, so they assume you run at 30Hz or higher. A constant tick rate (1/60 Hz EVERY frame gives the best stability). You CAN give it variable time-steps, but they shouldn't vary wildly.

A 1 second tick rate is far outside the acceptable range. The engine doesn't compute the full simulation during that time duration, it only computes where objects would be after 1 second and then sees whether objects collide with something at that position. Then it applies constraints so that objects don't overlap. In 1 second your objects can be anywhere, so they would tunnel through everything. You give it small timesteps so that the engine can see lots of steps in between and prevent that from happening.

And no, it won't kill parallelism, because the parallelism is not bound to the duration of the simulated timestep, but rather to the amount of objects that are active in your scene.

What most engines do is, if they get a timestep of 100ms, they execute 6 16ms physics timesteps and then save the remaining 4ms for later, so that when the next update comes they add it to that timestep. Optionally you can also do a variable timestep as long as it is below some threshold (say 32ms).

This is a complicated topic, and it depends a bit on how stable your simulation needs to be.

**Reply by TheFearCode** [reply]
I see, thank you.
However, can't you also achieve higher accuracy by diving your delta time into smaller fractions calling Update multiple times?
What I'm telling you is, I'm still struggling to understand the difference between:
`system.Update(2 / 60, 2, &_allocator, &_jobSystem);`
and
`system.Update(1/60, 1, &_allocator, &_jobSystem);
system.Update(1/60, 1, &_allocator, &_jobSystem);`
In my case, I apply variable timestep, I will start adopting iterative approach now.
But in the end, I see that the issue in my case is breaking the game's execution for debuging, that doesn't stop the system timer, so next time I compute delta it will include all the time the game was frozen, I need to not count that time somehow.

**Reply by jankrassnigg** [reply]
Every time you call update it also sends callbacks, eg for triggers and contacts. It does that once, even if you have 2 sub steps. So calling update once with more sub steps might be faster, but may not be what you want.
At least that's my understanding, Jorrit could probably elaborate better. But yeah, when you break in the debugger, you should not use wall clock time to continue. Instead advanced an internal timer by the difference, and clamp the difference to something like 100ms.

**Reply by jrouwe** [reply]
```system.Update(2 / 60, 2, &_allocator, &_jobSystem);``` should give the exact same simulation results as ```system.Update(1/60, 1, &_allocator, &_jobSystem); system.Update(1/60, 1, &_allocator, &_jobSystem);``` (including which and how many callbacks are called). The only reason why you would choose the former over the latter is that it is slightly faster because the job graph becomes a little bit more efficient. ```cMaxPhysicsJobs``` is high enough to do about 4 collision steps. If you need to do more than that you need to call Update multiple times with 4 steps per Update call (I don't think you would gain much speed by increasing the job count and number of steps further). 

I've never had to simulate more than 4/60 s since simulating that much time usually causes another frame drop and makes the situation only worse, often causing a spiral of death where it never catches up again. Just like @jankrassnigg, I slow down the clock in that case to allow the sim to catch up. So that boils down to clamping the world clock delta time to 66 ms.

**Reply by TheFearCode** [reply]
Thanks guys, that was very clarifying.
So, I'll keep the steps variable until 4, as suggested by @jrouwe , after that I'll start iterating the update call.
As for debugging, I'll clamp the timer (I was already clamping to 1 sec, I'll reduce it maybe to 66ms as suggested), although I only want to do that in debug so I may need to add yet another build flag to test network lags.

**Reply by jankrassnigg** [reply]
For debugger stuff, using this is often very handy: https://learn.microsoft.com/en-us/windows/win32/api/debugapi/nf-debugapi-isdebuggerpresent

Not sure if I would call it every frame, but you could call it on demand in case your timestep is very large, to determine whether you want to clamp it.

**Reply by TheFearCode** [reply]
Thanks for the suggestion, I'm aware about `IsDebuggerPresent` although that's only valid if we are in a debug build, in my case, that means DEBUG is defined, also I never run debug builds without a debugger and vice-versa, so I prefer to leave the evaluation at compile time to make sure I get a clean timestamp when not breaking the execution.
I'm making a PROFILER build configuration to not freeze the timer and be able to test the lag.
